### PR TITLE
[INSURANCE-67227] Added utility for vendors using payKonnect format

### PR DESCRIPTION
**Pull request by codewizzy**
JIRA: https://rippling.atlassian.net/browse/INSURANCE-67227

Added utility for vendors using payKonnect format.

### CODE REVIEW

File: flux_sdk/pension/utils/payKonnect.py [hunk 1]
```diff
@@ -0,0 +1,640 @@
...
+                        PayrollRunContribution
+                    ] = payroll_contribution_map.get(ContributionType.LOAN.name, None)
+
+                    payroll_date_absolute = (
+                        payroll_upload_settings.payrun_info.original_pay_date
+                        if hasattr(
+                            payroll_upload_settings.payrun_info, "original_pay_date"
+                        )
+                        and payroll_upload_settings.payrun_info.original_pay_date
+                        else None
```
Comment (stevevls) [inline flux_sdk/pension/utils/payKonnect.py:341]: Same as above w/ simplification to `getattr` with default.  I see more of them, so consider this comment a blanket statement. 😅

File: flux_sdk/pension/utils/payKonnect.py [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
+                    )
+                    payroll_date = (
+                        payroll_date_absolute.strftime("%m/%d/%Y")
```
Comment (stevevls) [inline flux_sdk/pension/utils/payKonnect.py:344]: Consider creating a constant for the date format string?

**Review by stevevls — COMMENTED** [review]
One high level suggestion: I think we should either lowercase  or snakecase `payKonnect` when it's part of a snakecase python name.  It looks really weird to me to see the mixed case in method and module names.

File: flux_sdk/pension/utils/payKonnect.py [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+        for payroll_contribution in payroll_contributions:
+            payroll_contribution_map[
+                payroll_contribution.deduction_type.name
+            ] = payroll_contribution
+
+        return payroll_contribution_map
+
+    @staticmethod
+    def _get_formatted_ssn(ssn: str) -> str:
+        return ssn[:3] + "-" + ssn[3:5] + "-" + ssn[5:9]
```
[the remark this answers is not in the document]
Comment (codewizzy) [reply flux_sdk/pension/utils/payKonnect.py:147]: Yes, that's the expected length, according to google "The SSN must be nine (9) numerical characters only."

**Comment by codewizzy** [comment]
--fix

File: flux_sdk/pension/utils/pay_konnect.py [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+
+        plan_name = payroll_upload_settings.customer_partner_settings.get(
+            "plan_id", "Plan_Number"
+        )
+
+        transmission_date = ReportPayrollContributionsPayKonnectUtil._get_today_date()
+        report_time = ReportPayrollContributionsPayKonnectUtil._pst_now().strftime(
+            "%H%M%S"
+        )
+        return "{}_{}_{}.csv".format(plan_name, transmission_date, report_time)
```
Comment (vguptarippling) [inline flux_sdk/pension/utils/pay_konnect.py:267]: Is this going to be the fixed format for all vendors ?

File: flux_sdk/pension/utils/pay_konnect.py [hunk 5]
```diff
@@ -0,0 +1,575 @@
...
+    @staticmethod
+    def _get_loa_info(leave_infos: list[LeaveInfo]) -> str:
+        leave_type = ""
+        for leave_info in leave_infos:
+            if (
+                leave_info.start_date
+                <= datetime.datetime.now().date()
+                <= leave_info.return_date
+            ):
+                leave_type = (
```
Comment (vguptarippling) [inline flux_sdk/pension/utils/pay_konnect.py:221]: Are we just considering the last leave here ? Is this communicated by PayKonnect ?

File: flux_sdk/pension/utils/pay_konnect.py [hunk 6]
```diff
@@ -0,0 +1,575 @@
...
+                        "Federal_Exemptions": "",
+                        "Projected Start Date": "",
+                        "Adjusted_Service_Date": "",
+                        "Match_Eligibility_Date": "",
+                        "Annual_Salary": annual_salary,
+                        "Hourly_Rate": "",
+                        "Earnings": "",
+                        "Earnings_YTD": "",
+                        "CONT_401K": employee_401k,
+                        "CONT_YTD_401K": "",
```
Comment (vguptarippling) [inline flux_sdk/pension/utils/pay_konnect.py:545]: Why are we not updating the ytd contribution for 401k ?

File: flux_sdk/pension/utils/pay_konnect.py [hunk 7]
```diff
@@ -0,0 +1,575 @@
...
+                        "CONT_Contribution_%_401K": "",
+                        "CONT_Contribution_$_401K": "",
+                        "CONT_Company_Match": "",
+                        "CONT_YTD_Company_Match": company_match_contribution,
```
Comment (vguptarippling) [inline flux_sdk/pension/utils/pay_konnect.py:549]: You are updating on incorrect column . This is for ytd_company_match not company_match. Please update it to CONT_Company_Match .
For ytd_updates. You can refer this codepiece https://github.com/Rippling/flux-apps/blob/main/flux_apps/charlesschwab/pension_kit/capabilities/report_payroll_contributions/implementation.py#L236

File: flux_sdk/pension/utils/pay_konnect.py (hunk 4 above)
Comment (codewizzy) [reply flux_sdk/pension/utils/pay_konnect.py:267]: I have asked this [qn](https://docs.google.com/spreadsheets/d/1u4ifxfC78rVZHqHCKyIvRtPyJXnL9erF6I2fepSFOT8/edit#gid=1988807008&range=B2) and waiting for response.
But according to Tim we don't need to be [blocked](https://rippling.slack.com/archives/C05GR5DRKV0/p1701948209189839?thread_ts=1701937842.340889&cid=C05GR5DRKV0) on these formatting queries and can update if required.

File: flux_sdk/pension/utils/pay_konnect.py (hunk 5 above)
Comment (codewizzy) [reply flux_sdk/pension/utils/pay_konnect.py:221]: This is to determine the current status of the employee, so using this we are determining if employee is currently on leave.

File: flux_sdk/pension/utils/pay_konnect.py (hunk 7 above)
Comment (codewizzy) [reply flux_sdk/pension/utils/pay_konnect.py:549]: Thanks for pointing out, updated.

File: flux_sdk/pension/utils/pay_konnect.py [hunk 8]
```diff
@@ -0,0 +1,577 @@
...
+                        "YTD_Total_Compensation": employee_year_to_date_gross_pay,
+                        "YTD_Plan_Compensation": "CLARIFY",
+                        "YTD_Excluded_Compensation": "",
+                        "Officer_Type": "",
+                        "Ownership_Percentage": "",
+                        "Highly_Comp_Code": "",
+                        "Participation_Date": "",
+                        "Eligibility_Code": "CLARIFY",
+                        "Salary_Amount": salary,
+                        "Termination_Reason_Code": "NEED TO ADD",
```
Comment (vguptarippling) [inline flux_sdk/pension/utils/pay_konnect.py:536]: Nit: can you instead use a todo for this ?

File: flux_sdk/pension/utils/pay_konnect.py [hunk 9]
```diff
@@ -0,0 +1,577 @@
...
+                        "Pay_Period_Excluded_Wages": "",
+                        "YTD_Hours_Worked": employee_year_to_date_hours_worked,
+                        "YTD_Total_Compensation": employee_year_to_date_gross_pay,
+                        "YTD_Plan_Compensation": "CLARIFY",
+                        "YTD_Excluded_Compensation": "",
+                        "Officer_Type": "",
+                        "Ownership_Percentage": "",
+                        "Highly_Comp_Code": "",
+                        "Participation_Date": "",
+                        "Eligibility_Code": "CLARIFY",
```
Comment (vguptarippling) [inline flux_sdk/pension/utils/pay_konnect.py:534]: Is this really a eligibility code ?

**Review by vguptarippling — APPROVED** [review]

File: flux_sdk/pension/utils/pay_konnect.py (hunk 9 above)
Comment (codewizzy) [reply flux_sdk/pension/utils/pay_konnect.py]: Going on discussion about this field

File: flux_sdk/pension/utils/pay_konnect.py [hunk 10]
```diff
@@ -0,0 +1,580 @@
...
+    def _get_payroll_contributions_map(
+        payroll_contributions: list[PayrollRunContribution],
+    ) -> dict[str, PayrollRunContribution]:
+        payroll_contribution_map = {}
+        for payroll_contribution in payroll_contributions:
+            payroll_contribution_map[
+                payroll_contribution.deduction_type.name
+            ] = payroll_contribution
+
+        return payroll_contribution_map
```
Comment (stevevls) [inline flux_sdk/pension/utils/pay_konnect.py:146]: this is unused now...let's delete

**Review by stevevls — APPROVED** [review]
Approving to unblock

**Pull request merged**